### PR TITLE
Add a paragraph describing the prohibition on implicit TCO.

### PR DESCRIPTION
**Pull request by sunfishcode**
This corresponds with https://github.com/WebAssembly/spec/pull/103

### CODE REVIEW

**Comment by sunfishcode** [comment]
Wordsmith integrated. FutureFeatures.md section added. I rewrote "likely to be added" with "planned for the future" which is stronger without quite suggesting more commitment from all parties involved than I'm comfortable making in this PR.

File: AstSemantics.md [hunk 1]
```diff
@@ -32,6 +32,18 @@ Callstack space is limited by unspecified and dynamically varying constraints
...
+Semantically, implementations must have an internal maximum call stack size,
+and every call must take up some resources toward exhausting that size. Of
+course, dynamic resources may be exhausted much earlier. This rule exists
+to avoid programs depending on tail call optimizations which may not be
+portable across implementations. Also, in the future, it is expected that
+WebAssembly will add some form of stack-introspection functionality, in which
+case such optimizations would be directly observable.
+
+Support for explicit tail calls is planned in
+[the future](FutureFeatures.md#general-purpose-proper-tail-calls).
```
Comment (lukewagner) [inline AstSemantics.md]: To address the above objection, perhaps append "which would add an explicit tail-call operation with well-defined effects on stack introspection."?

**Comment by sunfishcode** [comment]
Sounds good to me. Updated.

**Comment by jfbastien** [comment]
I discussed this with @titzer and @flagxor: we'd rather specify this positively instead of negatively. It would be better to say that the transform should eventually hit a limit, so an optimizer can do TCO but has to either prove that the corresponding loop terminates or keep an implementation-defined limit to model resource exhaustion (as detailed in Nondeterminism.md).

**Comment by sunfishcode** [comment]
@jfbastien If I remove the first sentence of the paragraph, does the rest satisfy your concern?

I added the first sentence as a "plain language" version of what follows, since the theoretical mechanism being inventing here is somewhat abstract and subtle, and I didn't want users or implementors to miss the main point. But I'm open to suggestions on how to clarify it.

**Comment by jfbastien** [comment]
@sunfishcode yes, dropping the first sentence is good.

I'd also clarify this one:
> This rule exists to avoid programs depending on tail call optimizations which may not be portable across implementations.

Calls are already not portable because of their implementation-defined limit! I'd instead say something like:
> This rule exists to avoid transforming a program which necessarily runs out of resources into one which cannot possibly exhaust any resource.

**Comment by sunfishcode** [comment]
@jfbastien Ok, I removed the first sentence and rewrote the other sentence. Does this work for you?

**Comment by jfbastien** [comment]
lgtm

**Comment by sunfishcode** [comment]
Merging with lgtm above and rough consensus in WebAssembly/spec#103.

**Pull request merged**